### PR TITLE
RUM-4633: Add support for custom TimeProvider

### PR DESCRIPTION
**Pull request by marco-saia-datadog**
### What does this PR do?

Implements an API function to allow the override of the default `TimeProvider`, used throughout RUM functions to retrieve the current timestamp.

### Additional Notes

- The current implementation handles the timestamp retrieval synchronously.

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)
- [ ] If this PR is auto-generated, please make sure also to manually update the code related to the change

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -36,6 +36,7 @@
         "@react-native/metro-config": "0.73.5",
         "@react-native/typescript-config": "0.73.1",
         "@testing-library/react-native": "7.0.2",
+        "@types/jest": "^29.5.12",
```
Comment (fuzzybinary) [inline package.json:39]: Was this a transitive dependency before?

File: packages/core/src/rum/__tests__/DdRum.test.ts [hunk 2]
```diff
@@ -16,9 +16,9 @@ import type { ErrorEventMapper } from '../eventMappers/errorEventMapper';
 import type { ResourceEventMapper } from '../eventMappers/resourceEventMapper';
 import { ErrorSource, PropagatorType, RumActionType } from '../types';
 
-jest.mock('../../utils/TimeProvider', () => {
+jest.mock('../../utils/time-provider/DefaultTimeProvider', () => {
     return {
-        TimeProvider: jest.fn().mockImplementation(() => {
+        DefaultTimeProvider: jest.fn().mockImplementation(() => {
```
Comment (fuzzybinary) [inline packages/core/src/rum/__tests__/DdRum.test.ts:21]: Why do we need a MockTimeProvider and a mock of the DefaultTimeProvider?

File: packages/core/src/rum/__tests__/DdRum.test.ts [hunk 3]
```diff
@@ -72,10 +77,10 @@ describe('DdRum', () => {
         });
 
         test('calls the native SDK when called with old API with default values', async () => {
-            await DdRum.startAction('scroll', 'page_old_api');
+            await DdRum.startAction(RumActionType.SCROLL, 'page_old_api');
             await DdRum.stopAction();
             expect(NativeModules.DdRum.stopAction).toHaveBeenCalledWith(
-                'scroll',
+                RumActionType.SCROLL,
```
Comment (fuzzybinary) [inline packages/core/src/rum/__tests__/DdRum.test.ts:83]: Should we do these replacements as part of this PR or separate it into its own PR?

**Review by fuzzybinary — APPROVED** [review]

File: package.json (hunk 1 above)
Comment (marco-saia-datadog) [reply package.json:39]: I believe it was just missing, making the IDE clueless about the correctness of the test files, so I simply added it in the dev dependencies.

File: packages/core/src/utils/Timer.ts [hunk 4]
```diff
@@ -15,10 +15,10 @@ const STOP_LABEL = '__stop';
  * All timestamps/durations returned are in milliseconds.
  */
 export default class Timer {
-    private timeProvider: TimeProvider;
+    private timeProvider: DefaultTimeProvider;
     private times: Record<string, Timestamp> = {};
 
-    constructor(timeProvider: TimeProvider = new TimeProvider()) {
+    constructor(timeProvider: DefaultTimeProvider = new DefaultTimeProvider()) {
```
Comment (0xnm) [inline packages/core/src/utils/Timer.ts]: shouldn't it be `TimeProvider` type here and `DefaultTimeProvider` as default value?

```suggestion
    constructor(timeProvider: TimeProvider = new DefaultTimeProvider()) {
```

**Review by 0xnm — APPROVED** [review]

File: packages/core/src/rum/__tests__/DdRum.test.ts (hunk 2 above)
Comment (marco-saia-datadog) [reply packages/core/src/rum/__tests__/DdRum.test.ts:21]: `DefaultTimeProvider` is the real use-case implementation of `TimeProvider` used as default inside `DdRum`. By mocking it in the tests we can check that the default implementation works as expected, when no custom time provider is set.

`MockTimeProvider` is a concrete implementation of `TimeProvider` that I have added in order to test the newly introduced `setTimeProvider` function.

File: packages/core/src/utils/Timer.ts (hunk 4 above)
Comment (marco-saia-datadog) [reply packages/core/src/utils/Timer.ts]: Of course it should, nice catch!

Comment (marco-saia-datadog) [reply packages/core/src/utils/Timer.ts]: [Solved](https://github.com/DataDog/dd-sdk-reactnative/compare/88c6563310ef6b04ad5009ae0c2444edcc810272..1ed72d363d95857bc4868bfa4df620b8e9a9fb63)

File: packages/core/src/rum/__tests__/DdRum.test.ts (hunk 3 above)
Comment (marco-saia-datadog) [reply packages/core/src/rum/__tests__/DdRum.test.ts:83]: I have thought about it, but I have decided to push it with this PR since these are pretty simple changes and they are in the scope of the newly implemented tests (kind of). Also, it's easy to verify that the tests still execute and pass correctly, so in my opinion we can keep it in this PR.
